<!DOCTYPE html>
<html lang="en">

<head>

    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, shrink-to-fit=no, initial-scale=1">
    <meta name="description" content="">
    <meta name="author" content="">

    <title>CoverView GUI</title>

    <!-- Bootstrap Core CSS -->
    <link href="static/css/bootstrap.min.css" rel="stylesheet">

    <!-- Custom CSS -->
    <link href="static/css/simple-sidebar.css" rel="stylesheet">

    <link href="static/css/tables.css" rel="stylesheet">
    <link href="static/css/radio.css" rel="stylesheet">

    <link href="static/css/buttons.css" rel="stylesheet">

    <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.1.1/jquery.min.js"></script>


</head>

<body>


<script type="text/javascript">


</script>


    <div id="wrapper">

        <!-- Sidebar -->
        <div id="sidebar-wrapper">
            <ul class="sidebar-nav">
                <li class="sidebar-brand"><a href="/">CoverView GUI</a></li>
                <li><a href="/analysis">Analysis View</a></li>
                <li onclick="goToDefaultGenesView();">Genes View</li>
                <li onclick="goToDefaultRegionsView();">Regions View</li>
                <li onclick="goToDefaultProfilesView();">Profiles View</li>
                <li id="thispage"><a href="/usage">Usage Guide</a></li>
            </ul>
        </div>
        <!-- /#sidebar-wrapper -->

        <!-- Page Content -->
        <div id="page-content-wrapper">
            <div class="container-fluid" style="position: relative; top: 10px">


                <div class="row">
                    <div class="col-lg-12">
                        <h1>Usage Guide</h1>
                        <br/>

                    </div>
                </div>

                <div class="row" style="position: relative; top: 20px;">
                    <div class="col-lg-12">
                        <h3>Analysis View</h3>
                        <p style="font-size: 18px;">A summary table of input files, key configuration options and the date of running the CoverView analysis. In addition, the number of <i>failed</i> regions and genes are also provided. (A gene is defined to be <i>failed</i> if it includes at least one <i>failed</i> region.)
                        If a text is too long to be displayed in the table, hover the mouse cursor over it to see the full text in the tooltip. </p>




                    </div>
                </div>

                <div class="row" style="position: relative; top: 20px;">
                    <div class="col-lg-12">
                        <h3>Genes View</h3>
                        <p style="font-size: 18px;">A bar plot displaying per-chromosome read counts and a table at the right side showing the total number of regions and the number of <i>failed</i> regions for each gene. The bar plot presents both on-target and off-target read counts for each chromosome. Clicking on a particular bar in the plot selects/unselects the chromosome.
                        If a chromosome is selected, only the genes located on that chromosome are listed in the table, otherwise all genes are shown.</p>

                            <p style="font-size: 18px;"> <i>Failed</i> genes (i.e. those with at least one <i>failed</i> region) are highlighted in red. The table can be filtered to show either <i>passed</i>, <i>failed</i> or all genes by the "Genes" option. The search box helps to find particular genes of interest.
                         </p>
                        <p style="font-size: 18px;">A gene can be selected/unselected by clicking on the row of the table. If a gene is selected, the "Show Regions" and "Show Failed Regions" buttons are visible. With "Show Regions" you can open the selected gene in Regions View where only the regions of the gene will be displayed. Clicking on the "Show Failed Regions" button also redirects to
                        Regions View, but only <i>failed</i> regions of the gene will be listed.</p>
                    </div>
                </div>

                <div class="row" style="position: relative; top: 20px;">
                    <div class="col-lg-12">
                        <h3>Regions View</h3>
                        <p style="font-size: 18px;">A scrollable table of the genomic regions that have been analysed by CoverView. Columns of the table provide the following region-specific metrics: RC, MEDCOV, MINCOV, MEDQCOV, MINQCOV, MAXFLBQ, MAXFLMQ and PASS. For <i>failed</i> regions, metrics values not satisfying quality requirements are highlighted in red.</p>

                        <p style="font-size: 18px;">The table can be filtered to show either <i>passed</i>, <i>failed</i> or all regions by using the "Regions" option. With the "Reads" option, you can switch between displaying the above metrics for forward reads only, reverse reads only or all reads included. The search box helps to find particular regions of interest.</p>
                        <p style="font-size: 18px;">A region can be selected/unselected by clicking on the row of the table. If a region is selected, the "Show Profiles" and "Show Gene" buttons are visible. With "Show Profiles" you can open the selected region in Profiles View. Clicking on the "Show Gene" button redirects to Gene View showing the gene to which
                        the selected region belongs.</p>

                    </div>
                </div>


                <div class="row" style="position: relative; top: 20px;">
                    <div class="col-lg-12">
                        <h3>Profiles View</h3>
                        <p style="font-size: 18px;">An interactive graph of per-base quality profiles aligned with the reference sequence. The visible profiles correspond to the region selected in the left-side panel.
                            The diplayed metrics can be changed with the selector at the top of both y-axes. The following per-base metrics can be selected: COV, QCOV, FLBQ, MEDBQ, FLMQ and MEDMQ. If two metrics are
                        plotted together, the color of each line is the same as of the corresponding metrics selector (blue for the y-axis and red for the y2-axis). Hover the mouse cursor over a point of the plot to see the exact genomic position and
                            metrics value.</p>
                        <p style="font-size: 18px;"> Click and drag on the plot to zoom in any section of the profiles. The genomic coordinates of the zoomed region are provided under the graph. You can move upstream and downstream or zoom out with
                            the navigation buttons. It is possible to repeatedly zoom in up to DNA base-level. If the "Axes" option is set to "Normal", the y-axes will have the same scale when you zoom in, otherwise they will be re-scaled to be optimised for the visible section of the profiles.</p>

                        <p style="font-size: 18px;"> With the "Reads" option, you can switch between showing the per-based profiles for forward reads only, reverse reads only or all reads included. Quality cut-off values used in the definition of
                        "PASS" requirements can be displayed as dashed horizontal lines by switching on the "Cut-off" option. A horizontal line representing the cut-off value is visible only if the relevant metrics is displayed and have the same
                        color.</p>

                        <p style="font-size: 18px;"> <i>Failed</i> regions are highlighted in red in the left-side panel. Clicking on the "Show Region" button will open the selected region in the Region View. </p>
                        <br/>
                        <br/>
                        <br/>
                        <br/>
                    </div>
                </div>



            </div>

        </div>

    </div>
    <!-- /#wrapper -->

    <script type="text/javascript" src="static/js/utils.js"></script>

    <!-- jQuery -->
    <script src="static/js/jquery.js"></script>

    <!-- Bootstrap Core JavaScript -->
    <script src="static/js/bootstrap.min.js"></script>

</body>

</html>
